### ISSUE TITLE
Invalid memory operation after running unittests

### ISSUE THREAD

**Issue by LiamM32** [LiamM32/Open_Emblem#20]
Since commit 271cf93, an exception is thrown after running unittests. Weirdly, it appears to happen *after* all unittests have passed. Weirdly, all this commit did was replace the `removeUnit` function in the `Map` class with a seemingly identical function through the `UnitArrayManagement` mixin template.

This function is called by the `Unit` class destructor, which is mentioned in the error output (specifically line 124 as of commit bf1c050). This means that the `Unit` destructor may be running *after* all the unittests have concluded. It may be a result of the `Map` or `Tile` object associated with a `Unit` being destroyed before the unit is destroyed.

Here is the unittest output on commit bf1c050:
```
             Generating test runner configuration 'open_emblem-test-regular' for 'regular' (sourceLibrary).
    Starting Performing "unittest" build using /usr/bin/dmd for x86_64.
  Up-to-date open_emblem ~master: target for configuration [open_emblem-test-regular] is up to date.
    Finished To force a rebuild of up-to-date targets, run again with --force
     Running open_emblem-test-regular 
Starting Vector2i unittest.
Vector2i unittest passed.
Starting Direction unittest.
Direction unittest passed.
Starting offsetByDirection unittest.
offsetByDirection unittest passed.
Starting `measureDistance` unittest.
30
`measureDistance` unittest passed.
Starting Map Unit deletion unittest.
Doing `removeUnit`
unitsArray: 
Unit A, 
Unit B, 
Unit C, 
shiftedUnits: 
Unit B, 
Unit C, 
shiftedUnits.length > 0
In loop. unitKey = 1, i = 0
Map unit deletion unittest passed.
Starting Map.checkObstruction unittest
`Map.checkObstruction` unittest passed.
Starting UnitStats unittest.
UnitStats unittest passed.
Starting Unit caching unittest.
Soldier location is now 5, 5
Updating tileReach for Soldier
Passed Unit caching unittest.
Doing `removeUnit`
unitsArray: 
shiftedUnits: 
Destroyed objects from Unit unittests.
3 modules passed unittests
Doing `removeUnit`
unitsArray: 
core.exception.InvalidMemoryOperationError@src/core/lifetime.d(126): Invalid memory operation
----------------
??:? onInvalidMemoryOperationError [0x561e42ab4018]
??:? nothrow void* core.internal.gc.impl.conservative.gc.ConservativeGC.runLocked!(core.internal.gc.impl.conservative.gc.ConservativeGC.mallocNoSync(ulong, uint, ref ulong, const(TypeInfo)), core.internal.gc.impl.conservative.gc.mallocTime, core.internal.gc.impl.conservative.gc.numMallocs, ulong, uint, ulong, const(TypeInfo)).runLocked(ref ulong, ref uint, ref ulong, ref const(TypeInfo)) [0x561e42aafa02]
??:? nothrow core.memory.BlkInfo_ core.internal.gc.impl.conservative.gc.ConservativeGC.qalloc(ulong, uint, scope const(TypeInfo)) [0x561e42aa74f1]
??:? gc_qalloc [0x561e42a746de]
??:? pure nothrow core.memory.BlkInfo_ rt.lifetime.__arrayAlloc(ulong, scope const(TypeInfo), const(TypeInfo)) [0x561e42a89477]
??:? _d_arraysetlengthiT [0x561e42a779a8]
/usr/include/dlang/dmd/core/internal/array/capacity.d:47 pure nothrow @trusted ulong core.internal.array.capacity._d_arraysetlengthTImpl!(immutable(char)[], immutable(char))._d_arraysetlengthT(return scope ref immutable(char)[], ulong) [0x561e42a3c0ff]
/usr/include/dlang/dmd/core/internal/array/concatenation.d:55 pure nothrow @trusted immutable(char)[] core.internal.array.concatenation._d_arraycatnTX!(immutable(char)[], immutable(char)[], immutable(char)[])._d_arraycatnTX(ref immutable(char)[], immutable(char)[]) [0x561e42a3c04c]
source/unit.d:405 bool map.Map.__mixin26.removeUnit(unit.Unit) [0x561e42a61822]
source/unit.d:124 scope void unit.Unit.__dtor() [0x561e42a67400]
??:? rt_finalize2 [0x561e42a76842]
/usr/include/dlang/dmd/object.d:4188 nothrow void object.destroy!(true, unit.Unit).destroy(unit.Unit) [0x561e42a3fb00]
source/map.d:128 scope void map.Map.__dtor() [0x561e42a631b3]
??:? rt_finalize2 [0x561e42a76842]
??:? rt_finalizeFromGC [0x561e42ab34c6]
??:? nothrow ulong core.internal.gc.impl.conservative.gc.Gcx.sweep() [0x561e42aaaae9]
??:? nothrow ulong core.internal.gc.impl.conservative.gc.Gcx.fullcollect(bool, bool, bool) [0x561e42aabcd6]
??:? nothrow ulong core.internal.gc.impl.conservative.gc.ConservativeGC.runLocked!(core.internal.gc.impl.conservative.gc.ConservativeGC.fullCollectNoStack().go(core.internal.gc.impl.conservative.gc.Gcx*), core.internal.gc.impl.conservative.gc.Gcx*).runLocked(ref core.internal.gc.impl.conservative.gc.Gcx*) [0x561e42ab051f]
??:? nothrow void core.internal.gc.impl.conservative.gc.ConservativeGC.fullCollectNoStack() [0x561e42aa8763]
??:? nothrow void core.internal.gc.impl.conservative.gc.ConservativeGC.collectNoStack() [0x561e42aa86f1]
??:? gc_term [0x561e42a9f903]
??:? rt_term [0x561e42a88b1e]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x561e42a757b0]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x561e42a75699]
??:? _d_run_main2 [0x561e42a75602]
??:? _d_run_main [0x561e42a753eb]
/usr/include/dlang/dmd/core/internal/entrypoint.d:29 main [0x561e42a3a8f1]
??:? [0x7f12b46d6ccf]
??:? __libc_start_main [0x7f12b46d6d89]
??:? _start [0x561e42a3a7d4]
Error Program exited with code 1
```

**Comment by LiamM32** [comment]
I found that the error shown above doesn't happen if the block of `writeln` lines in `unit.UnitArrayManagement.removeUnit` is removed. It then instead gives a simple `Error Program exited with code -11`. However, the first error shouldn't happen, as that would mean that it's trying to call a function for an array entry that should have already been deleted. When running GDB, it also points to the Unit destructor for where the segmentation fault happens.

I should also note that for most of these recent commits, I had forgotten to replace all mentions of the `MapTemp!(Tile,Unit)` with `Map`, which means that you must do the edits yourself in order to replicate my results with a particular commit.

**Comment by LiamM32** [comment]
Solved in 64109e5.
